<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of compositional data • CompoData</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">CompoData</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/CompoData.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Analysis of compositional data</h1>
                        <h4 class="author">Emanuel Heitlinger</h4>
            
            <h4 class="date">2018-02-14</h4>
          </div>

    
    
<div class="contents">
<div id="analysis-of-compositional-count-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis-of-compositional-count-data" class="anchor"></a>Analysis of compositional count data</h1>
<div id="e-g-in-microbiomes-or-transcriptomes" class="section level2">
<h2 class="hasAnchor">
<a href="#e-g-in-microbiomes-or-transcriptomes" class="anchor"></a>(e.g. in microbiomes or transcriptomes)</h2>
<p>Standard analyses of abundance estimates genereated from sequencing treat the generated data as counts. Methods for differential abundance estimation in RNAseq data have probably recieved most attention in this regard (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">see e.g. this for one of the current approaches</a>). Many (if not all) of the methods developed for analyses of count data derived from transcriptomes are also applied to amplicon sequencing data. On the other hand techniques from classical community ecolgy are applied on data from transcriptomics or amplicon sequencing alike (mostly after <em>transforming</em> it to count data).</p>
<p>In the field of amplicon sequencing there are currently concernes that analysis of “compositional data” as counts can introduce biases (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5695134/">e.g. reviewd here</a>, refering in its core to <a href="https://www.amazon.de/Statistical-Compositional-Monographs-Statistics-Probability/dp/0412280604">Aitchison, J. (1986)</a>).</p>
<p>This document aims to give an introduction to the challenges that are posed by compositional data produced to assess transcriptomes via RNAseq or microbiomes via amplicon sequencing. I will give an “abstract” introduction into:</p>
<ol style="list-style-type: decimal">
<li>pecularities of “compositional count data” from sequencing,</li>
</ol>
<p>and show</p>
<ol start="2" style="list-style-type: decimal">
<li>how analyses of community ecology (try to) deal with these (or not).</li>
</ol>
<p>The main purpose is to discuss with the audience how concepts in community ecology and data analyses are affected by the nature of sequencing data and I want to discuss</p>
<ol start="3" style="list-style-type: decimal">
<li>whether it is even conceptually possible to address these issues using generic solutions.</li>
</ol>
<p>This is distributed and “developed” as a package to ease further expansion and contribution.</p>
</div>
<div id="so-what-is-this-compositional-data" class="section level2">
<h2 class="hasAnchor">
<a href="#so-what-is-this-compositional-data" class="anchor"></a>So what is this “compositional data”?</h2>
<p>Imagine you would count animals in two different ecosystems (e.g. on mars, earth and venus) but at the end forget your overall count. What you keep is just the relative proporion of tigers, unicorns and ladybugs.</p>
<p>Why should such a strange thing happen?</p>
<p>When sequencing DNA fragments, we sample a pool of molecules. We usually lack the information on the number of entities (organisms or transcripts) this pool of molecues was generated from (<a href="https://www.nature.com/articles/nature24460">but see this as a counter example</a>).</p>
<p>To illustrate we start to simulate some count data (strongly inspired by <code><a href="http://www.rdocumentation.org/packages/DESeq2/topics/makeExampleDESeqDataSet">DESeq2::makeExampleDESeqDataSet</a></code>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(reshape)
## initialize the random number generator (-&gt; reproducibility)
<span class="kw">set.seed</span>(<span class="dv">1209</span>)

n &lt;-<span class="st"> </span><span class="dv">3</span> ## number of rows
m &lt;-<span class="st"> </span><span class="dv">3</span> ## number of columns

means &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">8</span>, <span class="dv">28</span>, <span class="dv">330</span>) ## per row (species) mean

countData &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnbinom</span>(m *<span class="st"> </span>n, <span class="dt">mu =</span> means, <span class="dt">size =</span> <span class="dv">10</span>), 
                    <span class="dt">ncol =</span> m)

<span class="kw">colnames</span>(countData) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"venus"</span>, <span class="st">"earth"</span>, <span class="st">"mars"</span>)
<span class="kw">rownames</span>(countData) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"tigers"</span>, <span class="st">"unicorns"</span>, <span class="st">"ladybugs"</span>)

<span class="kw">ggplot</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/reshape/topics/melt-24">melt</a></span>(countData),
       <span class="kw">aes</span>(<span class="dt">x=</span>X2, <span class="dt">y=</span>value, <span class="dt">fill=</span>X1)) +
<span class="st">    </span><span class="kw">geom_col</span>()</code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-1-1.png" width="384"></p>
<p>These values (given how we simulated them) could now be the actual counts of our animals on those planets. The problem is: when sequencing we don’t know. PCRs in the preparation of libraries (collections of DNA molecules) perserve the relative proportions of different molecules (say those from tigers, unicorns and ladybugs) but provide no information on the number before amplification.</p>
<p>In addition, our sampling of the finally produced molecules usually varies in depth (that is sequencing depth varies). A usual first step in all analyses is to account for this variation in sampling depths. Standard (presented below) methods don’t consider the compositional nature of sequencing data when dong so.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">depth &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="fl">0.1</span>, <span class="fl">0.5</span>) ## per column (planet) depth
mu &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(means)%*%<span class="kw">t</span>(<span class="kw">as.matrix</span>(depth))

countData &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnbinom</span>(m *<span class="st"> </span>n, <span class="dt">mu =</span> mu, <span class="dt">size =</span> <span class="dv">10</span>), 
                   <span class="dt">ncol =</span> m)
<span class="kw">colnames</span>(countData) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"venus"</span>, <span class="st">"earth"</span>, <span class="st">"mars"</span>)
<span class="kw">rownames</span>(countData) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"tigers"</span>, <span class="st">"unicorns"</span>, <span class="st">"ladybugs"</span>)

<span class="kw">ggplot</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/reshape/topics/melt-24">melt</a></span>(countData),
       <span class="kw">aes</span>(<span class="dt">x=</span>X2, <span class="dt">y=</span>value, <span class="dt">fill=</span>X1)) +
<span class="st">    </span><span class="kw">geom_col</span>()</code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-2-1.png" width="384"></p>
</div>
<div id="scaling-by-totals-1" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-by-totals-1" class="anchor"></a>Scaling by totals (1)</h2>
<p>The simplest method to adjust for the depth of sequencing is to scale by the number of total reads per sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scaling.factor &lt;-<span class="st"> </span><span class="kw">colSums</span>(countData)/<span class="kw">max</span>(<span class="kw">colSums</span>(countData))

<span class="kw">ggplot</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/reshape/topics/melt-24">melt</a></span>(<span class="kw">t</span>(<span class="kw">t</span>(countData)*<span class="dv">1</span>/scaling.factor)),
       <span class="kw">aes</span>(<span class="dt">x=</span>X2, <span class="dt">y=</span>value, <span class="dt">fill=</span>X1)) +
<span class="st">    </span><span class="kw">geom_col</span>()</code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-3-1.png" width="384"></p>
</div>
<div id="interlude-a-more-realistic-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#interlude-a-more-realistic-dataset" class="anchor"></a>Interlude: A more realistic dataset</h2>
<p>We now adjust a couple of things in our dataset to make it more realistic:</p>
<ol style="list-style-type: decimal">
<li><p>We add more species (rows, could also be genes in trascriptomics)</p></li>
<li><p>We add more samples (columns) and also specify them as replicates for our differnt ecosystmes (could be tissues or treatments in transcriptomics)</p></li>
<li><p>Differences between ecosystems</p></li>
<li><p>We make the dispersion (negative binomial parameter, a.k.a. parameter of the gamma distribution describing the variation in the poisson paramter of a composite poisson-gamma) vary with the mean of the count for a species.</p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(pheatmap)
n &lt;-<span class="st"> </span><span class="dv">300</span> ## number of rows (species)
m &lt;-<span class="st"> </span><span class="dv">36</span> ## number of columns (samples)

## the relationship between mean and dispersion
dispMeanRel &lt;-<span class="st"> </span>function(x) <span class="dv">4</span>/x +<span class="st"> </span><span class="fl">0.1</span>

## low mean and high standard deviation to get sparse counts (lots of
## zeros)
interceptMean &lt;-<span class="st"> </span><span class="fl">0.1</span>
interceptSD &lt;-<span class="st"> </span><span class="dv">2</span>

## the depth as influenced by our sequencing
## depth &lt;- runif(m, min=0.4, max=1)

##  for now assume we have even depths
depth &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span>, <span class="dt">times=</span>m)

## beta will be the mean of the negative binomial it is here still on
## the log scale we make two more columns wich give a deviation from
## our intercept SD (see below)
beta &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">rnorm</span>(n, interceptMean, interceptSD),
              <span class="kw">rnorm</span>(n, <span class="dv">0</span>, <span class="dv">1</span>),
              <span class="kw">rnorm</span>(n, <span class="dv">0</span>, <span class="dv">2</span>))

## the sample different ecosystems
colData &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">planet =</span>
                          <span class="kw">factor</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">"venus"</span>, <span class="st">"earth"</span>, <span class="st">"mars"</span>),
                                     <span class="dt">each =</span> m/<span class="dv">3</span>)))
<span class="kw">rownames</span>(colData) &lt;-<span class="st"> </span><span class="kw">paste</span>(colData$planet, <span class="dv">1</span>:m, <span class="dt">sep=</span><span class="st">"_"</span>)

## a model matrix telling us which sample belongs to which condition
planet.design &lt;-<span class="st"> </span>stats::<span class="kw">model.matrix.default</span>(~colData$planet)

## the final sample means based on design, log-means and the depth of
## sequencing. By matrix multiplication mars gets 
mu &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="dv">2</span>^(planet.design %*%<span class="st"> </span><span class="kw">t</span>(beta)) *<span class="st"> </span>depth)

## the dispersion via the mean relationship 
dispersion &lt;-<span class="st"> </span><span class="kw">dispMeanRel</span>(<span class="dv">2</span>^(beta[, <span class="dv">1</span>]))

## and based of on these the negative binomial counts
countData &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnbinom</span>(m *<span class="st"> </span>n, <span class="dt">mu =</span> mu, <span class="dt">size =</span> dispersion), 
                    <span class="dt">ncol =</span> m)

## we now remove species for which we have only zero counts (we
## wouldn't have this in a real dataset)
countData &lt;-<span class="st"> </span>countData[<span class="kw">rowSums</span>(countData)&gt;<span class="dv">0</span>, ]

<span class="kw">colnames</span>(countData) &lt;-<span class="st"> </span><span class="kw">rownames</span>(colData)
<span class="kw">rownames</span>(countData) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"species"</span>, <span class="dv">1</span>:<span class="kw">nrow</span>(countData), <span class="dt">sep=</span><span class="st">"_"</span>)

<span class="kw"><a href="http://www.rdocumentation.org/packages/pheatmap/topics/pheatmap">pheatmap</a></span>(<span class="kw">log10</span>(countData<span class="dv">+1</span>),
         <span class="dt">annotation_col=</span>colData,
         <span class="dt">show_rownames=</span><span class="ot">FALSE</span>,
         <span class="dt">show_colnames=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-4-1.png" width="384"></p>
<p>I aimed for a simulation of a dataset which resembles more amplicon sequencing as compred to RNAseq. These datasets are usually sparse, meaning many species have in many samples zero counts. Mean values of a negative binomial distribution for individual species are themselfs normally distributed. These mean values are low but have a high variance. To differenciate the different ecosystems I used differences in the mean which are normally distibuted with a mean of zero.</p>
</div>
<div id="scaling-by-totals-2-slightly-more-realistic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#scaling-by-totals-2-slightly-more-realistic-data" class="anchor"></a>Scaling by totals (2, slightly more realistic data)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
ColS &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample.sum=</span><span class="kw">colSums</span>(countData), <span class="dt">num=</span><span class="dv">1</span>:<span class="kw">ncol</span>(countData),
                   <span class="dt">planet=</span>colData$planet)

<span class="kw">ggplot</span>(ColS, <span class="kw">aes</span>(<span class="dt">x=</span><span class="kw">reorder</span>(num, sample.sum), <span class="dt">y=</span>sample.sum, <span class="dt">fill=</span>planet))+
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>)

scaling.factor &lt;-<span class="st"> </span><span class="kw">colSums</span>(countData)/<span class="kw">max</span>(<span class="kw">colSums</span>(countData))

countData.scaled &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(countData)*<span class="dv">1</span>/scaling.factor)

<span class="kw"><a href="http://www.rdocumentation.org/packages/pheatmap/topics/pheatmap">pheatmap</a></span>(<span class="kw">log10</span>(countData.scaled<span class="dv">+1</span>),
         <span class="dt">annotation_col=</span>colData,
         <span class="dt">show_rownames=</span><span class="ot">FALSE</span>,
         <span class="dt">show_colnames=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-5-1.png" width="384"><img src="CompoData_files/figure-html/unnamed-chunk-5-2.png" width="384"></p>
<p>This lead to a dataset in which the total number of animals varied a lot. A situation which should be challenging when the data is “normalized”.</p>
<p>Interestingly, scaleing by totals seems to screw us over here. We did not include sequencing depth differnences in this simulation, so we know we wouldn’t have needed to scale. But annoyingly the sequencing process itself (or to be more exact the PCR) does “scale” for us. In real data we don’t know all of this and we have to correct for the differences in sequencing depths.</p>
</div>
<div id="rarefaction" class="section level2">
<h2 class="hasAnchor">
<a href="#rarefaction" class="anchor"></a>Rarefaction</h2>
<p>What if we wanted to estimate species richness? For any kind of basic species count or for the richness indices we need to correct for our observation effort!</p>
<p>This means we need to sub-sample every sample to the number of sequencing reads we obtained for the least deeply assessed sample we want to include. We do this repeatedly to get an estimate of the variance this induces.</p>
<p>We use a the <code>vegan</code> function to perform this rarefication. Ecologists traditionally have put the species in the columns and sites in the rows, probably as they had more sites than spcies. We have in most cases more species than sites and do it the other way round (at least I do).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(vegan)
<span class="co">#&gt; Loading required package: permute</span>
<span class="co">#&gt; Loading required package: lattice</span>
<span class="co">#&gt; This is vegan 2.4-6</span>
countData.rare &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span>:<span class="dv">100</span>, function (i){
    <span class="kw">t</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/vegan/topics/rarefy">rrarefy</a></span>(<span class="kw">t</span>(countData),
              <span class="dt">sample=</span><span class="kw">min</span>(ColS$sample.sum)))
})

<span class="kw"><a href="http://www.rdocumentation.org/packages/pheatmap/topics/pheatmap">pheatmap</a></span>(<span class="kw">log10</span>(countData.rare[[<span class="dv">1</span>]]+<span class="dv">1</span>),
         <span class="dt">annotation_col=</span>colData,
         <span class="dt">show_rownames=</span><span class="ot">FALSE</span>,
         <span class="dt">show_colnames=</span><span class="ot">FALSE</span>)

richness &lt;-<span class="st"> </span><span class="kw">lapply</span>(countData.rare, function(y){
    <span class="kw">apply</span>(y, <span class="dv">2</span>, function(x) <span class="kw">length</span>(x[x&gt;<span class="dv">0</span>]))
})

rich.mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">unlist</span>(richness), <span class="dt">nrow=</span>m)

richness &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">planet=</span>colData$planet,
                       <span class="dt">richness.mean=</span><span class="kw">rowMeans</span>(rich.mat),
                       <span class="dt">richness.sd=</span><span class="kw">apply</span>(rich.mat, <span class="dv">1</span>, sd),
                       <span class="dt">n=</span><span class="dv">1</span>:<span class="kw">nrow</span>(rich.mat))

<span class="kw">ggplot</span>(richness, <span class="kw">aes</span>(<span class="kw">as.character</span>(n), richness.mean)) +
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">2</span>) +
<span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin=</span>richness.mean-richness.sd,
                      <span class="dt">ymax=</span>richness.mean+richness.sd))+
<span class="st">    </span><span class="kw">facet_wrap</span>(~planet, <span class="dt">scales=</span><span class="st">"free_x"</span>)</code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-6-1.png" width="384"><img src="CompoData_files/figure-html/unnamed-chunk-6-2.png" width="384"></p>
<p>Uh, we are screwed again!</p>
</div>
<div id="using-totals-based-scaling-as-offset-for-modelling" class="section level2">
<h2 class="hasAnchor">
<a href="#using-totals-based-scaling-as-offset-for-modelling" class="anchor"></a>Using totals based scaling as offset for modelling</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(DESeq2)

<span class="kw">mode</span>(countData) &lt;-<span class="st"> "integer"</span>

rowRanges &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="st">"1"</span>, <span class="kw">IRanges</span>(<span class="dt">start =</span> (<span class="dv">1</span>:<span class="kw">nrow</span>(countData) -<span class="st"> </span><span class="dv">1</span>) *<span class="st"> </span><span class="dv">100</span> +<span class="st"> </span><span class="dv">1</span>,
                                  <span class="dt">width =</span> <span class="dv">100</span>))

dds &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/DESeqDataSet">DESeqDataSetFromMatrix</a></span>(<span class="dt">countData =</span> countData, <span class="dt">colData =</span> colData, 
                              <span class="dt">design =</span> ~planet, <span class="dt">rowRanges =</span> rowRanges,
                              <span class="dt">ignoreRank =</span> <span class="ot">TRUE</span>)

dds &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/estimateSizeFactors">estimateSizeFactors</a></span>(dds, <span class="dt">type=</span><span class="st">"poscounts"</span>)

dds &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/estimateDispersions">estimateDispersions</a></span>(dds)
<span class="co">#&gt; gene-wise dispersion estimates</span>
<span class="co">#&gt; mean-dispersion relationship</span>
<span class="co">#&gt; -- note: fitType='parametric', but the dispersion trend was not well captured by the</span>
<span class="co">#&gt;    function: y = a/x + b, and a local regression fit was automatically substituted.</span>
<span class="co">#&gt;    specify fitType='local' or 'mean' to avoid this message next time.</span>
<span class="co">#&gt; final dispersion estimates</span>
dds &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/nbinomLRT">nbinomLRT</a></span>(dds, <span class="dt">reduced=</span> ~<span class="dv">1</span>)
res.deseq &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/DESeq2/topics/results">results</a></span>(dds)

<span class="kw">head</span>(res.deseq[<span class="kw">order</span>(res.deseq$padj), ])
<span class="co">#&gt; log2 fold change (MLE): planet venus vs earth </span>
<span class="co">#&gt; LRT p-value: '~ planet' vs '~ 1' </span>
<span class="co">#&gt; DataFrame with 6 rows and 6 columns</span>
<span class="co">#&gt;              baseMean log2FoldChange     lfcSE      stat       pvalue</span>
<span class="co">#&gt;             &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;</span>
<span class="co">#&gt; species_213 94.702455       5.857367 0.6183568 142.25641 1.286512e-31</span>
<span class="co">#&gt; species_28   3.072630       4.375525 0.6857651 105.53118 1.213936e-23</span>
<span class="co">#&gt; species_21   5.300482       3.474154 0.5985473  72.24978 2.047194e-16</span>
<span class="co">#&gt; species_55   7.374460       3.088628 0.5440076  70.40334 5.153587e-16</span>
<span class="co">#&gt; species_74   4.125340      -4.206402 0.7922348  65.15906 7.093993e-15</span>
<span class="co">#&gt; species_237  8.332600       3.240092 0.5555765  64.47082 1.000783e-14</span>
<span class="co">#&gt;                     padj</span>
<span class="co">#&gt;                &lt;numeric&gt;</span>
<span class="co">#&gt; species_213 3.293471e-29</span>
<span class="co">#&gt; species_28  1.553838e-21</span>
<span class="co">#&gt; species_21  1.746939e-14</span>
<span class="co">#&gt; species_55  3.298295e-14</span>
<span class="co">#&gt; species_74  3.632124e-13</span>
<span class="co">#&gt; species_237 4.270006e-13</span>
<span class="kw">table</span>(res.deseq$padj&lt;<span class="fl">0.01</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;   150   106</span></code></pre></div>
</div>
<div id="techniques-acknowleding-compositionality" class="section level2">
<h2 class="hasAnchor">
<a href="#techniques-acknowleding-compositionality" class="anchor"></a>Techniques acknowleding compositionality</h2>
<div id="classical-testing-for-differential-abundance" class="section level3">
<h3 class="hasAnchor">
<a href="#classical-testing-for-differential-abundance" class="anchor"></a>Classical testing for differential abundance</h3>
<p>Centered log-ratio (clr) transformation offer an alternative (<a href="https://github.com/ggloor/CoDa_microbiome_tutorial/wiki/Before-we-start">see here</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ALDEx2)

g.mean &lt;-<span class="st"> </span>function (x, <span class="dt">na.rm=</span><span class="ot">TRUE</span>){
    <span class="kw">exp</span>(<span class="kw">sum</span>(<span class="kw">log</span>(x[x &gt;<span class="st"> </span><span class="dv">0</span>]), <span class="dt">na.rm=</span>na.rm) /<span class="st"> </span><span class="kw">length</span>(x))
}

countData.clr &lt;-<span class="st"> </span><span class="kw">apply</span>(countData, <span class="dv">2</span>, function(x){
    <span class="kw">log</span>(x<span class="fl">+0.0001</span>/<span class="kw">g.mean</span>(x<span class="fl">+0.0001</span>))
})

<span class="kw"><a href="http://www.rdocumentation.org/packages/pheatmap/topics/pheatmap">pheatmap</a></span>(countData.clr,
         <span class="dt">annotation_col=</span>colData,
         <span class="dt">show_rownames=</span><span class="ot">FALSE</span>,
         <span class="dt">show_colnames=</span><span class="ot">FALSE</span>)

countData.clrP &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ALDEx2/topics/aldex.clr.function-methods">aldex.clr</a></span>(countData,
                            <span class="dt">conds=</span><span class="kw">as.character</span>(colData$planet))
<span class="co">#&gt; [1] "operating in serial mode"</span>
<span class="co">#&gt; [1] "computing center with all features"</span>

res.aldex &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ALDEx2/topics/aldex.glm">aldex.glm</a></span>(countData.clrP, <span class="dt">conditions=</span>colData$planet)
<span class="co">#&gt; [1] "operating in serial mode"</span>
<span class="co">#&gt; [1] "running tests for each MC instance:"</span>
<span class="co">#&gt; |------------(25%)----------(50%)----------(75%)----------|</span></code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-8-1.png" width="384"></p>
<p>The problem here is that the neither the geometric mean nor the log can be calculated for zero-count samples. After our naive solution (adding a small number) we used a package which “converts the single estimate to a probability vector”. This uses Monte-Carlo sampling from the Dirichlet distribution (<a href="https://doi.org/10.17713/ajs.v45i4.122">ref</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">merge</span>(res.aldex, <span class="kw">as.data.frame</span>(res.deseq), <span class="dt">by=</span><span class="dv">0</span>)

<span class="kw">ggplot</span>(res, <span class="kw">aes</span>(glm.eBH, padj)) +
<span class="st">    </span><span class="kw">geom_point</span>() +
<span class="st">    </span><span class="kw">scale_y_log10</span>() +
<span class="st">    </span><span class="kw">scale_x_log10</span>()
<span class="co">#&gt; Warning: Removed 41 rows containing missing values (geom_point).</span></code></pre></div>
<p><img src="CompoData_files/figure-html/unnamed-chunk-9-1.png" width="384"></p>
<p>The p-values from very established negative-binomial glm (using scaling as an offset parameter) are highly corrlated. The compositional approach is a bit more conservative in the multiple-testing adjusted p-values.</p>
</div>
<div id="so-what-could-it-be-good-for" class="section level3">
<h3 class="hasAnchor">
<a href="#so-what-could-it-be-good-for" class="anchor"></a>So what could it be good for</h3>
<p>MY guess: correlation between samples</p>
<blockquote>
<p>To be continued.</p>
</blockquote>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#analysis-of-compositional-count-data">Analysis of compositional count data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#e-g-in-microbiomes-or-transcriptomes">(e.g. in microbiomes or transcriptomes)</a></li>
      <li><a href="#so-what-is-this-compositional-data">So what is this “compositional data”?</a></li>
      <li><a href="#scaling-by-totals-1">Scaling by totals (1)</a></li>
      <li><a href="#interlude-a-more-realistic-dataset">Interlude: A more realistic dataset</a></li>
      <li><a href="#scaling-by-totals-2-slightly-more-realistic-data">Scaling by totals (2, slightly more realistic data)</a></li>
      <li><a href="#rarefaction">Rarefaction</a></li>
      <li><a href="#using-totals-based-scaling-as-offset-for-modelling">Using totals based scaling as offset for modelling</a></li>
      <li><a href="#techniques-acknowleding-compositionality">Techniques acknowleding compositionality</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
